Create custom Selectors (Kqueue, Epoll, Poll, and Select) Backport

SethMichaelLarson opened the pull request:
This PR adds all the `select` methods to a new util module called `wait`. Uses the same ordering used by `selectors.py` in the standard library to determine which selector should be used. Also factored in a way to allow easy patching and testing. Similar to #998 but uses a different way to break up the different methods to wait for I/O events. Also adds tests to exercise the selectors which can probably be extended.

Lukasa commented:
@SethMichaelLarson Rather than us doing this, would it be better simply to vendor Python 3.5's `selectors` module? Or, possibly, just have `selectors` as an optional dependency for using something other than `select`? I'm a bit inclined to want to take a general solution that someone else built rather than write our own custom code that is more likely to be partial or incomplete.

Thoughts?

SethMichaelLarson commented:
@Lukasa This is what asyncio does, while possible I think it may be overkill in this situation as a selector has much more overhead compared to a polling object. Asyncio has a different use-case for those selectors to what we're using them for.

Lukasa commented:
@SethMichaelLarson Do we have a good feel on what the overhead is?

SethMichaelLarson commented:
@Lukasa Probably minimal but I can take a look if you're interested. I'm +1 on this suggestion. In the future if we get to doing your suggestion of having IOCP being used for selecting on Windows we'll have to have our own separate module as well but for now this is the best solution.

Lukasa commented:
Heh, if we get to the point of doing IOCP we're going to be in a whole separate world of pain anyway. ;)

SethMichaelLarson commented:
So I'll open a separate PR or just keep it in this one?

Lukasa commented:
Let's keep the work here and just retitle the PR a bit.

SethMichaelLarson commented:
Sure thing, will re-title when I commit.

SethMichaelLarson commented:
So a lot of talk has happened on this issue but took place in issue #1000 and I'm going to summarize this here.

Basically there is no easy option. The `selectors` standard library module does exactly what we want but is only available in Python 3.5+. The `selectors34` module supports Python 2 and Python 3 but does not have the same functionality as `selectors` for Python 3.5 which is to restart the select in the case of a system interrupt which is the new behavior for system calls specified by [PEP 475](https://www.python.org/dev/peps/pep-0475/). We're left with no good options but to make this to be product level code in order to support this functionality. Due to difficulties with getting a CI with Solaris, @Lukasa and I have agreed to drop `devpoll` support in favor of using `epoll` on Solaris (`devpoll` is Solaris only) and `kqueue` is fine because it may be available on Mac OS. This will require setting up Mac OS on Travis as well.

The basic requirements for this issue is now:
- [ ] Start with `selectors` and add functionality.
- [ ] Remove `devpoll` support.
- [ ] Catch `EINTR` errors and restart the syscall.
- [ ] Create full test suite for this code.
- [ ] Maintain 100% test coverage after adding Mac OS to Travis.

Did I miss anything, @Lukasa ?

Lukasa commented:
Nope, looks good. By the time we get to the last bullet we should come to a position on whether we want to use something like coveralls for test coverage across multiple platforms, but let's do one step at a time.

SethMichaelLarson commented:
I'd recommend codecov over coveralls, just better functionality and integration with Github. /bikeshed :wink: But honestly, being able for it to comment on PRs is nice.

Lukasa commented:
Let's not put the cart before the horse. ;) Let's get kqueue support *first*, and then worry about testing it. =D

SethMichaelLarson commented:
@Lukasa Agreed!

SethMichaelLarson commented:
So I've done a little work and implemented `SelectSelector` to start us off as it's the simplest.  The biggest part of commit [`983a193`](https://github.com/shazow/urllib3/pull/1001/commits/983a1939e6286249e213be9f53b68c128b98e0af) is the `_syscall_wrapper` function. This function is made to wrap system calls and to retry them if they fail in the middle of executing due to an `EINTR` error. Usage of this function is required within all system calls for this module as Python 3.4< do not handle this themselves, instead returning an empty list.

From this starting point it will be adding a test suite for all selectors and implementing `KqueueSelector`, `EpollSelector` and `PollSelector` which are all much more syscall intense than `SelectSelector`.  I'd like a review on this code before progressing further.

Lukasa commented on urllib3/packages/selectors.py:
```diff
@@ -0,0 +1,288 @@
+if hasattr(time, "monotime"):
+    monotonic = time.monotonic
+else:
+    monotonic = time.time
```
Rather than do `if hasattr...else`, better to do `try...except AttributeError`.

SethMichaelLarson replied:
Sure, also noticed a mistake using "monotime" instead of "monotonic".

Lukasa commented on urllib3/packages/selectors.py:
```diff
@@ -0,0 +1,288 @@
+    result = None
+    while expires is None:
+        current_time = monotonic()
+        if current_time > expires:
```
With `timeout=0` this may fail immediately, because `expires` is calculated based off an earlier call to `monotonic()`. We should probably allow one run with `timeout=0`, so you may want to have `current_time` updated once before this loop begins (and use it to calculate `expires`), and then update at the *bottom* of the loop.

Lukasa replied:
I should note as well that this guarantees that `expires` is `None`, and on Python 3 comparing a number to `None` raises a `TypeError`.

SethMichaelLarson replied:
I will fix this check, the next commit cycle will include tests so I can catch these trivial errors.

Lukasa commented on urllib3/packages/selectors.py:
```diff
@@ -0,0 +1,288 @@
+        expires = monotonic() + timeout
+
+    result = None
+    while expires is None:
```
As noted below, this loop condition *cannot* possibly be right. Presumably you mean `while result is None`.

Lukasa replied:
However, this doesn't allow for the possibility that the *syscall* may return `None`. Now, I think that's unlikely, but to be defensive against that possibility let's create a sentinel object that we use instead of `None`.

SethMichaelLarson replied:
Wow, complete mistype, can't believe I missed that one. I'll change that. And yes we can use a sentinel object just in case a syscall returns None.

Lukasa commented on urllib3/packages/selectors.py:
```diff
@@ -0,0 +1,288 @@
+        self.close()
+
+
+class _BaseSelectorImpl(BaseSelector):
```
Given that we don't need the utility of an ABC (and indeed given that `BaseSelector` *is not* an ABC in this codebase), do we need this separation between `BaseSelector` and `_BaseSelectorImpl`?

SethMichaelLarson replied:
I was wondering this myself, probably not. I'll merge the two classes. This first revision was mostly to get the basics from selectors down.

Lukasa commented on urllib3/packages/selectors.py:
```diff
@@ -0,0 +1,288 @@
+            if events & EVENT_READ:
+                self._readers.add(key.fd)
+            if events & EVENT_WRITE:
+                self._readers.add(key.fd)
```
This looks wrong.

Lukasa commented on urllib3/packages/selectors.py:
```diff
@@ -0,0 +1,288 @@
+elif 'EpollSelector' in globals():
+    DefaultSelector = EpollSelector
+elif 'PollSelector' in globals():
+    DefaultSelector = PollSelector
```
Obviously all of this jazz raises `NameError`s all over the place.

SethMichaelLarson replied:
Shouldn't, this is the same method of doing things that's used in `selectors.py`. I can change to something else though.

SethMichaelLarson commented:
Thanks for the initial review @lukasa

SethMichaelLarson commented:
@lukasa Bit of information that I gleaned from PEP 475:
`"(Note: the selector module already retries on InterruptedError , but it doesn't recompute the timeout yet)"` Apparently the `selector` module should (?) be recomputed the timeout, currently with this implementation the same timeout is used on a retry but it would be made to use a timeout based on the initial timeout amount minus how much time was spent selecting before the interrupt. We should probably take a position on this.

I'm in favor of recalculating the timeout time for certain system calls, perhaps have a parameter called `recalc_timeout` in `_syscall_wrapper` that will recalculate timeout if it's passed via `kwargs`. Reason being that timeout times will be regular rather than randomly being twice as long based on a system interrupt.

SethMichaelLarson commented:
So I added a lot of tests (inspired by  `asyncio`'s `selectors` tests) and I've got coverage up to 94% on Python3.5, the only area that isn't covered by a test is the system call catcher, but that'll get covered by Travis. I've also implemented the `PollSelector`, so now only `EpollSelector` and `KqueueSelector` remain.

Would appreciate another review at this point, especially if all tests pass on Travis.

SethMichaelLarson commented:
So I've done a lot of work on this, I've got all tests passing for `SelectSelector`, `PollSelector`, and `EpollSelector`, full coverage on all of those selectors, but now we're arriving at more interesting problems that require collaboration with a maintainer.

The first is that Kqueue cannot be assessed for coverage because it is not implemented on the Travis Linux machines. @Lukasa brought up Mac OS being used for testing this selector in addition to a code coverage service such as coveralls or codecov in order to bridge the platform gap in coverage reports.

The second interesting problem is that the `unittest` module doesn't support `unittest.skipUnless` until Python 2.7+, but it is used in order to check whether the platform supports selectors before doing unittests. Wanted to ask here for ideas on how to resolve this issue.

I've updated the checklist above as well.

Lukasa commented:
Ok, let's start by adding macOS and a coverage tool to Travis. For various boring reasons relating to permissions it's probably best if I do that myself directly and we merge it to master, and then you do a catch-up merge to grab the new travis.yml and whatever.

As to the unittest module, the easiest thing to do is going to be to add a test dependency on the unittest2 package for Python 2.6, and then do a conditional import as needed. Does that sound sensible?

Lukasa commented:
Ok, the adding of OS X is progressing in #1002.

SethMichaelLarson commented:
Sounds good to me, that's what I figured would have to happen. :) I'll add that conditional test dependency and sync with upstream when the Mac OS is running. Thanks!

SethMichaelLarson commented:
I've updated the checklist and split some items up.

SethMichaelLarson commented:
Also: feel free to cancel any builds from this PR if they're blocking your own builds, I've got my own Travis instance running as well. :)

SethMichaelLarson commented:
So I've got the additional `unittest2` installation happening for Python 2.6 in tox and it seems to be working on the Linux builds. I also added a few stress tests, one from `asyncio` and the other is my own creation. Both attempt to push the selectors to high usage and see what breaks.

I'll take some time to write some Windows specific tests as well, even though we can't test them via CI it'll help me feel better about releasing this code because of the strange limits Windows puts on `select` when writing code from my Windows machine.

SethMichaelLarson commented:
Now that the codecov and Mac OS builds are online I can progress on this. :) Thanks @lukasa !

SethMichaelLarson commented:
So another comment, when I get home today I'm going to add in a "shiv" for `socket.socketpair` so that tests can be run on Windows with Python <3.5 similar to how `asyncio` does things.

SethMichaelLarson commented:
Haha, to be honest working on these types of things are my reprieve from my actual job. Don't get me wrong, I do plenty of other things when I get a break, but I really do enjoy this low-level interactions with tons of platform differences. No matter how frustrating it can be, it's fun to finally build something that works and you're proud of. :)

Also, I'm glad that this work enables this project to go even farther! I'm excited to know what the future holds and hopefully I'll be of assistance then as well. :)

SethMichaelLarson commented:
Also unrelated but taking a guess at what your idea is, does it perhaps have something to do with enhancing the efficiency of ConnectionPools?

Lukasa commented:
Not right now it doesn't, though it certainly could in future.

SethMichaelLarson commented:
Because I'm insane I'm also going to tag here that I'm going to experiment with IOCP for Windows as well, I can't guarantee anything because I'm not very fluent with the Windows API and in all honesty if it doesn't work out right now it can be added later.  Only reason I'm attempting this at all is because asyncio and a few other references I've found laying around give a pretty good overview of the relevant parts of the Windows API I would be using.  Plus I've got a nice Windows box to work and test all this stuff on, so it won't be just stabbing in the dark. :)

SethMichaelLarson commented:
So it looks like the `KqueueSelector` is making things difficult, I'm removing the multi-event test for now until I can get a handle on what exactly is being returned by the internal `kqueue` object.  I'll do this on a private branch for experimentation so as not to cloud this PR or Travis up with commits.

SethMichaelLarson commented:
Travis failed build for Python 2.7 intermittently, could someone kick that build?

SethMichaelLarson commented:
Hitting 100% according to codecov now! :) Ready for review @Lukasa!

Lukasa commented:
Awesome @SethMichaelLarson! I'm not guaranteeing to code review this until Monday (I'm currently on a long mission to preserve my weekends as much as possible), but in the meantime: can one or both of @glyph and @hawkowl confirm Seth's belief about kqueue? It doesn't seem right analytically,  but I'd like to be sure.

glyph commented:
I do not think that is accurate; kqueue should be perfectly happy to have read & write selected at the same time.  In Twisted, we do achieve this with two separate calls to `kq.control` rather than masking out `KQ_FILTER_READ` and `KQ_FILTER_WRITE` in the same call, though; I don't know if that makes a difference.

Lukasa commented:
It did seem unlikely to me. @SethMichaelLarson want to pursue that train of thought a bit more?

SethMichaelLarson commented:
I've been looking at it for 1.5 days now, `KQ_FILTER_READ` and `KQ_FILTER_WRITE` aren't bit masks, they're just negative numbers so they're not meant to be combined and using `KQ_EV_ADD` twice seems to only give back the last one you added to the kqueue when you actually go to look for events. I've done a ton of searching on the docs, StackOverflow, Google, all other selector implementations. 

The thing is, other selectors implementations aren't trying to do this, none of them have a test in place for watching an fd for both reading and writing.  They don't support multiple events being watch on the same fd, otherwise a copy-paste solution of selectors' `KqueueSelector` would pass the `test_multiple_event_types`, but it doesn't. Works for all other selectors though, but that's because their event types are bit masks and can be combined.

I welcome any help and I will continue looking.

SethMichaelLarson commented:
Figured out the problem, it wasn't anything to do with the `KqueueSelector` implementation. I didn't have a sleep in the test. :disappointed: Hopefully that'll pass with full coverage now.

SethMichaelLarson commented:
And now it looks like one of the Mac OS builders locked up on installing Python 3.4 :disappointed:

Lukasa commented:
Restarted

SethMichaelLarson commented:
Looks like it passed! :) Ready for review on Monday. Thanks for pushing me to try just a tiny bit more haha @Lukasa  Also @Lukasa are you going to turn on the 'required' 100% codecov PR check now that we're using codecov?

Lukasa commented:
I'm going to require @shazow's help with that. I'll chase it up soonish. =)

SethMichaelLarson commented:
@Lukasa Sure thing! Enjoy your weekend! :)

shazow commented:
How can I help? :D

shazow commented on urllib3/util/selectors.py:
```diff
@@ -0,0 +1,544 @@
+# Backport of selectors.py from Python 3.5+ to support Python < 3.4
```
Should this live along urllib3/packages/backports/ or somesuch?

Lukasa replied:
For the moment I'd rather it lives here, I think. Given that this is actually a fairly customised backport that we are trying to enforce code coverage on, I think we need to consider it part of our product rather than something that we just trust to work.

SethMichaelLarson commented:
@shazow We just need your help setting codecov to be a required check to merge because we've dropped the old coverage check with nosetests in favor of codecov so we can do cross-platform coverage checks.

Lukasa commented:
Yup: I think codecov needs to be activated for this repo, and I can't do it. 😎

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,622 @@
+        self.assertEqual(key.fileobj, rd)
+        self.assertEqual(key.fd, rd.fileno())
+        self.assertEqual(key.events, selectors.EVENT_READ)
+        self.assertEqual(key.data, "data")
```
We probably want to pass something we can assert identity on, rather than asserting equality.

SethMichaelLarson replied:
I can fix this.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,622 @@
+        wr.close()
+
+        s.unregister(rdfd)
+        s.unregister(wrfd)
```
This test and the one below it don't actually make any assertions. They should be refactored so that they do.

SethMichaelLarson replied:
Sure thing.

SethMichaelLarson replied:
Now checks that the map is actually empty.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,622 @@
+            wr.send(b'x')
+
+        # Give time to flush the writes.
+        time.sleep(1.0)
```
Too long. =D

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,622 @@
+        self.assertLess(monotonic() - t, 0.1)
+
+        t = monotonic()
+        self.assertEqual(0, len(s.select(timeout=1)))
```
Too long.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,622 @@
+        self.addCleanup(signal.alarm, 0)
+
+        # Start the timer for the interrupt.
+        signal.alarm(1)
```
Too long.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,622 @@
+        self.addCleanup(signal.alarm, 0)
+
+        # Start the timer for the interrupt.
+        signal.alarm(1)
```
Too long.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,622 @@
+        self.addCleanup(signal.alarm, 0)
+
+        # Start the timer for the interrupt.
+        signal.alarm(1)
```
Too long.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,622 @@
+
+        t = monotonic()
+        self.assertEqual([(key, selectors.EVENT_READ)], s.select(2))
+        self.assertLess(monotonic() - t, 2.2)
```
As noted above we probably don't need this assert.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,622 @@
+        self.addCleanup(signal.alarm, 0)
+
+        # Start the timer for the interrupt.
+        signal.alarm(1)
```
Too long.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,622 @@
+
+        def first_alarm(*args):
+            signal.alarm(0)
+            signal.alarm(1)
```
Also too long.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,622 @@
+    # specification from PEP 475 which is that any system call
+    # that is interrupted and the interrupt handler raises an
+    # exception should not retry the system call and instead
+    # raise the exception.
```
This paragraph is not quite grammatically accurate. =P

SethMichaelLarson replied:
I will fix this.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,622 @@
+        s.close()
+        after_fds = len(proc.open_files())
+        if before_fds != after_fds:
+            self.skipTest("Leaked fd but test was on Windows.")
```
This is a bit weird. Do we gain much from this test? We can't really assert the result of it, so it's not really telling us all that much.

SethMichaelLarson replied:
Only reason I added it is to test on Windows, skipping being a "failure" to the developer.

Lukasa replied:
I don't think that we can rely on that. Developers won't spot that "skip means fail".

SethMichaelLarson replied:
Sure, I will remove this test.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,622 @@
+        if before_fds != after_fds:
+            self.skipTest("Leaked fd but test was on Windows.")
+
+class ScalableSelectorMixin:
```
Can we please explicitly inherit from `object`?

SethMichaelLarson replied:
Probably, I don't think Mixins typically do but I'll do this.

Lukasa replied:
I think they have to, actually, because otherwise in Python 2 they count as an old-style class and everything gets really sad really fast.

Lukasa replied:
I mean, logically, on Python 3 anything that doesn't explicitly inherit from `object` still *implicitly* inherits from `object`, so one way or another all mixins inherit from `object`.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,622 @@
+class ScalableSelectorMixin:
+    """ Mixin to test selectors that allow more fds than FD_SETSIZE """
+    @skipUnless(resource, "Could not import the resource module")
+    @skipUnless(sys.platform != "darwin", "Can't run on Mac OS due to RINFINITE hard limit.")
```
I feel like `skipIf` is going to be clearer with this specific check.

SethMichaelLarson replied:
Yeah it would be.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,622 @@
+
+        # NOTE: AttributeError resource.RLIMIT_OFILE is not defined on Mac OS.
+        except (OSError, resource.error, AttributeError):
+            pass
```
Given that we don't test on macOS, we presumably don't need this block.

SethMichaelLarson replied:
We do now! :)

Lukasa replied:
Yup, this test is skipped on Darwin systems.

SethMichaelLarson replied:
Ah, I forgot to change that to something else. So in the actual selectors they test on some Mac OS systems not all, I need to change that so we do the same. Then there will be a need for this.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,622 @@
+                # Too many FDs should skip. *BSD and Solaris fail on
+                # connecting or binding rather than on create.
+                if e.errno == errno.EMFILE or e.errno == errno.EADDRNOTAVAIL:
+                    self.skipTest("RLIMIT_NOFILE limit reached.")
```
Once again I don't love that we skip the test in this case. I'd argue it should probably be a failure mode.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,622 @@
+                # connecting or binding rather than on create.
+                if e.errno == errno.EMFILE or e.errno == errno.EADDRNOTAVAIL:
+                    self.skipTest("RLIMIT_NOFILE limit reached.")
+                    break
```
This is an entirely superfluous `break` statement.

SethMichaelLarson replied:
I added it so my IDE doesn't complain about some warning.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,622 @@
+                if e.errno == errno.ENOSPC:
+                    # This can be raised by epoll if we go
+                    # over fs.epoll.max_user_watches sysctl
+                    self.skipTest("MAX_USER_WATCHES reached.")
```
Again, skipping tests in this case feels weird.

Lukasa commented on test/test_util.py:
```diff
@@ -40,6 +41,8 @@
 from . import clear_warnings
 
+from nose.plugins.skip import SkipTest
+
```
You have a few rogue imports in this file.

Lukasa commented on urllib3/contrib/pyopenssl.py:
```diff
@@ -465,7 +461,7 @@ def ssl_wrap_socket(sock, keyfile=None, certfile=None, cert_reqs=None,
             cnx.do_handshake()
         except OpenSSL.SSL.WantReadError:
-            rd, _, _ = select.select([sock], [], [], sock.gettimeout())
+            rd = util.wait_for_read(sock, sock.gettimeout())
```
Longer term we should re-evaluate this. We should try to avoid the overhead of repeatedly creating selectors if we can.

SethMichaelLarson replied:
That would certainly speed things up.

Lukasa commented on urllib3/util/connection.py:
```diff
@@ -26,22 +23,13 @@ def is_connection_dropped(conn):  # Platform-specific
-            # Either data is buffered (bad), or the connection is dropped.
-            return True
+    try:
+        return wait_for_read(sock, timeout=0.0)
```
Can we force-cast this to a boolean please?

Lukasa commented on urllib3/util/connection.py:
```diff
@@ -26,22 +23,13 @@ def is_connection_dropped(conn):  # Platform-specific
-            return True
+    try:
+        return wait_for_read(sock, timeout=0.0)
+    except select.error:
```
It might be nicer to provide a custom exception we can track, rather than piggybacking on the stdlib here.

Lukasa commented on urllib3/util/connection.py:
```diff
@@ -92,7 +80,6 @@ def create_connection(address, timeout=socket._GLOBAL_DEFAULT_TIMEOUT,
             err = e
             if sock is not None:
                 sock.close()
-                sock = None
```
Where has this gone?

SethMichaelLarson replied:
sock gets set to None at the top of the loop regardless of if it's set at the bottom. I will restore it.

Lukasa commented on urllib3/util/selectors.py:
```diff
@@ -0,0 +1,544 @@
+
+    args = list(args)
+    if recalc_timeout and "timeout" not in kwargs and not (
+                len(args) and isinstance(args[-1], (float, int))):
```
Do we need to support passing the timeout via positional argument?

SethMichaelLarson replied:
`select.select` doesn't take a positional argument, but if needed I can always wrap it like I did with `select.poll`.

SethMichaelLarson replied:
I think I will wrap it in a similar way so we can only do positional timeouts. :)

Lukasa commented on urllib3/util/selectors.py:
```diff
@@ -0,0 +1,544 @@
+    while result is _SYSCALL_SENTINEL:
+        try:
+            result = func(*args, **kwargs)
+            break
```
This `break` statement is redundant with the `_SYSCALL_SENTINEL`. Either the `while` should be changed to `while True` or the `break` should be removed.

SethMichaelLarson replied:
I'll remove the break.

Lukasa commented on urllib3/util/selectors.py:
```diff
@@ -0,0 +1,544 @@
+                        else:
+                            args[-1] = expires - current_time
+                continue
+            raise
```
I think we might prefer to wrap exceptions here. I believe I noted this above.

SethMichaelLarson replied:
Sure, what name would we like to have for this exception?

Lukasa replied:
Good question. `SelectorError`, maybe?

SethMichaelLarson replied:
Sure I will make this change.

Lukasa commented on urllib3/util/selectors.py:
```diff
@@ -0,0 +1,544 @@
+        raise NotImplementedError()
+
+    def close(self):
+        """ Close the selector. This must be called to insure that all
```
`ensure`.

Lukasa commented on urllib3/util/selectors.py:
```diff
@@ -0,0 +1,544 @@
+            return key
+
+        def select(self, timeout=None):
+
```
What's this empty line for?

SethMichaelLarson replied:
Nothing. :P I'll remove.

Lukasa commented on urllib3/util/selectors.py:
```diff
@@ -0,0 +1,544 @@
+        def select(self, timeout=None):
+
+            # Selecting on empty lists on Windows errors out.
+            if not (self._readers | self._writers):
```
Is a set union a good idea here, or do we just want to do boolean or?

SethMichaelLarson replied:
We can just do boolean or. I will change this.

Lukasa commented on urllib3/util/selectors.py:
```diff
@@ -0,0 +1,544 @@
+            ready = []
+            r, w, _ = _syscall_wrapper(select.select, timeout, True,
+                                       self._readers, self._writers,
+                                       [], timeout)
```
It's really weird that we pass timeout twice here.

SethMichaelLarson replied:
It is a bit. I can try to refactor it to only use the kwargs timeout along with my comment above.

SethMichaelLarson commented on test/test_selectors.py:
```diff
@@ -0,0 +1,622 @@
+
+        t = monotonic()
+        self.assertEqual([rd], selectors.wait_for_read(rd, timeout=2.0))
+        self.assertLess(monotonic() - t, 2.2)
```
I can remove this assert.

SethMichaelLarson commented on test/test_selectors.py:
```diff
@@ -0,0 +1,622 @@
+from urllib3.util import selectors
+
+
+@skipUnless(selectors.HAS_SELECT, "Platform doesn't have a selector and socketpair")
```
Not any more, used to be. I will remove.

SethMichaelLarson commented on urllib3/util/selectors.py:
```diff
@@ -0,0 +1,544 @@
+                    _syscall_wrapper(self._kqueue.control, None,
+                                     False, [kevent], 0, 0)
+                except OSError:
+                    pass
```
It's in unregister because interacting with stale/closed fds can cause problems and we don't want to fail on unregistering an fd that shouldn't be registered.

SethMichaelLarson commented:
Thank you @Lukasa for the extensive review! :) I've replied to a few of the reviews with questions and will be implementing changes hopefully later today.

SethMichaelLarson commented:
@Lukasa Something to note about the long tests is that `signal.alarm` only takes integer arguments, so instead I will have to perhaps make a threaded solution to an alarm that can take a sub-second amount of time to throw a signal. I will think about this more.

SethMichaelLarson commented:
Random CI fail on Mac OS. Hopefully will resolve when I address the rest of the comments. :pray:

SethMichaelLarson commented:
@Lukasa So I addressed almost all the changes in the code and if the change wasn't made I left a comment on the review. Ready for the next round of reviewing. :) Thanks for keeping this feedback loop so quick, I really appreciate it! It's my favorite part of waking up early. :)

SethMichaelLarson commented:
So MacOS + Python 3.[345] failed for some random reason that I don't understand, maybe you know why it failed @lukasa? Maybe restart them?

SethMichaelLarson commented:
Also codecov is showing up as a required check!!! :hooray:

SethMichaelLarson commented:
Random fail on Mac OS Python 2.6, possible restart?

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,670 @@
+        selectors.wait_for_read([rd], timeout=0.01)
+        t = get_time() - t
+        self.assertGreater(t, 0.0)
+        self.assertLess(t, 0.1)
```
Can we package these asserts up into a helper assert we have in a base class? That'd be super convenient. Maybe even in a helper context manager that *also* manages the calls to `get_time`?

SethMichaelLarson replied:
Yeah I can do this.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,670 @@
+            sel.register(wr, selectors.EVENT_WRITE)
+
+        self.assertRaises(RuntimeError, s.get_key, rd)
+        self.assertRaises(RuntimeError, s.get_key, wr)
```
We may want to assert that `get_key` *does* return what we expect inside the context manager.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,670 @@
+        os.dup2(wr2.fileno(), wrfd)
+
+        s.unregister(rdfd)
+        s.unregister(wrfd)
```
No assertions here.

SethMichaelLarson replied:
I will add asserts

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,670 @@
+        wr.close()
+
+        s.unregister(rd)
+        s.unregister(wr)
```
We still have no assertions here.

SethMichaelLarson replied:
Ah how do I miss this. :/

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,670 @@
+        if hasattr(s, "fileno"):
+            fd = s.fileno()
+            self.assertTrue(isinstance(fd, int))
+            self.assertGreaterEqual(fd, 0)
```
This test probably *should* formally skip itself if the selector does not have a `fileno` attribute.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,670 @@
+            if tuple(map(int, platform.mac_ver()[0].split("."))) < (major, minor):
+                return False
+        except ValueError:
+            pass
```
Why are we suppressing errors here?

SethMichaelLarson replied:
I actually don't know, the asyncio/test_selectors has this test and doesn't say anything about why the ValueError is there. Should I remove it and see what happens?

Lukasa replied:
Yup. =)

Lukasa commented on test/test_util.py:
```diff
@@ -544,4 +544,4 @@ def test_parse_retry_after(self):
         self.assertEqual(retry.parse_retry_after("0"), 0)
         self.assertEqual(retry.parse_retry_after("1000"), 1000)
-        self.assertEqual(retry.parse_retry_after("\t42 "), 42)
+        self.assertEqual(retry.parse_retry_after("\t42 "), 42)
```
This missing newline will make @haikuginger sad. 😎

SethMichaelLarson replied:
That's true and we can't have that! :)

SethMichaelLarson commented:
Btw did you notice that codecov is showing up now? :)

Lukasa commented:
I did! Hopefully this means that @shazow can now add it as a required check. =D

SethMichaelLarson commented:
Oh @Lukasa, I also was wondering if it would be worth publishing this selectors file as it's own module because `selectors34` only works for 3.4+ and this in theory should work for 2.6+.  If I am allowed to publish it on my own I would like to, I won't claim copyright or license it under anything but MIT or any of those things but if you are worried about those things I will refrain from publishing.

Lukasa commented:
You are absolutely free to republish the selectors code from this module as a standalone package. You can and should publish it with a license compatible with this module and with the PSF license, but you should also include yourself in the list of copyright owners as you've done a big chunk of the work. The easiest thing to do is to publish under the same license as this repo, and then include both the license for urllib3 and the license for Python 3.5 in a NOTICES file.

I hope others get some use out of this backport. =D

SethMichaelLarson commented:
@Lukasa Thank you so much, I will do this when this merge is complete so that we know that everything is working as intended. :)

SethMichaelLarson commented:
@Lukasa Didn't say it yesterday but this is ready for re-review. :)

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,684 @@
+
+        with self.assertTakesTime(upper=LONG_SELECT):
+            ready = s.select(LONG_SELECT)
+            self.assertEqual([(key, selectors.EVENT_READ)], ready)
```
As a basic note about cleanliness, you should probably remove the actual assertion from the context manager in this kind of case.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,684 @@
+        # Start the timer for the interrupt.
+        self.set_alarm(SHORT_SELECT)
+
+        with self.assertTakesTime(upper=LONG_SELECT):
```
This should presumably be `upper=SHORT_SELECT`?

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,684 @@
+        # Start the timer for the interrupt.
+        self.set_alarm(SHORT_SELECT)
+
+        with self.assertTakesTime(lower=SHORT_SELECT * 2, upper=LONG_SELECT):
```
Once again, we presumably want to *fail* this test if it takes as long as `LONG_SELECT`, but this context manager would allow that.

SethMichaelLarson replied:
I will change this.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,684 @@
+        except selectors.SelectorError as e:
+            self.assertEqual(e.errno, 100)
+        except Exception:
+            self.fail("Raised incorrect exception")
```
Let's show what the exception is.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,684 @@
+        # Start the timer for the interrupt.
+        self.set_alarm(SHORT_SELECT)
+
+        with self.assertTakesTime(upper=LONG_SELECT):
```
Do we actually expect this to take `LONG_SELECT` time?

SethMichaelLarson replied:
No you are totally correct this should only take `SHORT_SELECT` time.  Good catch.

Lukasa replied:
This is still `LONG_SELECT`.

Lukasa commented on urllib3/util/selectors.py:
```diff
@@ -0,0 +1,558 @@
+def _fileobj_to_fd(fileobj):
+    """ Return a file descriptor from a file object. If
+    given an integer will simply return that integer back. """
+    if isinstance(fileobj, six.integer_types):
```
Is `integer_types` really necessary here? Can a Python 2 long integer be a FD?

SethMichaelLarson replied:
Don't think so, I will use int.

Lukasa commented on urllib3/util/selectors.py:
```diff
@@ -0,0 +1,558 @@
+            is_interrupt = (errcode == errno.EINTR or (hasattr(errno, "WSAEINTR") and
+                                                       errcode == errno.WSAEINTR))
+
+            if errcode is not None and is_interrupt:
```
Why does this have an `errcode is not None` stanza? Can `is_interrupt` be true if `errcode` is `None`?

Lukasa commented on urllib3/util/selectors.py:
```diff
@@ -0,0 +1,558 @@
+class BaseSelector(object):
+    """ Abstract Selector class
+
+    A select supports registering file objects to be monitored
```
`selector`

SethMichaelLarson replied:
I will fix this.

Lukasa commented on urllib3/util/selectors.py:
```diff
@@ -0,0 +1,558 @@
+    for specific I/O events.
+
+    A file object is a file descriptor or any object with a
+    `fileno()` method. An arbitrary object can be attaced to the
```
`attached`

Lukasa commented on urllib3/util/selectors.py:
```diff
@@ -0,0 +1,558 @@
+        """ Return the key associated with a registered file object. """
+        mapping = self.get_map()
+        if mapping is None:
+            raise RuntimeError("Selector is closed")
```
It's infuriating that this raises a `RuntimeError`, but I accept that we have to do this because the standard library does.

SethMichaelLarson replied:
Yes, I do not understand why they chose RuntimeError but this is what was chosen so I chose it as well.

SethMichaelLarson commented on test/test_selectors.py:
```diff
@@ -0,0 +1,684 @@
+        rd, wr = socketpair()
+        self.addCleanup(rd.close)
+        self.addCleanup(wr.close)
+        return rd, wr
```
Selectors will work either way but I will make them non-blocking. :)

SethMichaelLarson commented on test/test_selectors.py:
```diff
@@ -0,0 +1,684 @@
+        # Start the timer for the interrupt.
+        self.set_alarm(SHORT_SELECT)
+
+        with self.assertTakesTime(lower=LONG_SELECT):
```
I guess that's true, the only reason I didn't set it is because I feared having random failures on the CI. I will try it and see what happens.

Lukasa replied:
Basically we want our tests to be as accurate as possible. While random failures on the CI are possible, it just means our tolerances aren't high enough. :grin:

SethMichaelLarson replied:
That is very true, I will make this change. :)

SethMichaelLarson commented on test/test_selectors.py:
```diff
@@ -0,0 +1,684 @@
+        self.addCleanup(signal.signal, signal.SIGALRM, sigalrm_handler)
+
+        # Start the timer for the interrupt.
+        self.set_alarm(SHORT_SELECT)
```
Sure thing, I can add it to the `AlarmMixin`.

Lukasa replied:
So you added the helper to the mixin but didn't actually use it here. :wink:

SethMichaelLarson commented on test/test_selectors.py:
```diff
@@ -0,0 +1,684 @@
+        self.assertEqual(rd.recv(1), b'x')
+
+
+@skipUnless(selectors.HAS_SELECT, "Platform doesn't have a selector and socketpair")
```
Gah, sorry. I will remove.

SethMichaelLarson commented on test/test_selectors.py:
```diff
@@ -0,0 +1,684 @@
+        # Start the timer for the interrupt.
+        self.set_alarm(SHORT_SELECT)
+
+        with self.assertTakesTime(lower=SHORT_SELECT, upper=LONG_SELECT):
```
Sure, I will do this. I'm glad this makes reviewing easier. :)

SethMichaelLarson commented on test/test_selectors.py:
```diff
@@ -0,0 +1,684 @@
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+        s.register(rd, selectors.EVENT_READ, "data_rd")
+        s.register(wr, selectors.EVENT_WRITE, "data_wr")
```
Sure, I will remove these.

SethMichaelLarson commented on test/test_selectors.py:
```diff
@@ -0,0 +1,684 @@
+        key = s.register(rd, selectors.EVENT_READ, d1)
+        key2 = s.modify(rd, selectors.EVENT_READ, d2)
+        self.assertEqual(key.events, key2.events)
+        self.assertNotEqual(key.data, key2.data)
```
Good idea I will add this.

SethMichaelLarson commented:
Thank you @Lukasa for the great review. :) I know this is taking a long time but with each iteration I feel better about this product. :)

SethMichaelLarson commented:
Also I don't know if anyone else is experiencing this, but the load time of this page is in the multi-second range now. This PR is getting long. ;)

Lukasa commented:
The more cycles of review we do the better I feel about this kind of thing. There's no way a reviewer can hope to really assess as patch this large and complex, so giving myself lots of opportunities to swing back to it means that we get much better *effective* coverage. So I'm pleased that you're willing to do this long back-and-forth.

And, yeah, this is becoming quite the patch! I'm not sure it's the one with the largest overhead that I've worked on yet, but it's definitely getting there. =D Let's keep it up!

SethMichaelLarson commented:
@Lukasa Yes I am always eager to receive my Github notification with your review, this PR has become a part of my daily routine. :)

SethMichaelLarson commented:
I believe I've hit all changes with this commit, I will look through them for a third time. :) Let's hope that the tighter constraints don't start failing on the CI, they don't fail on my machine but Travis is notably slower and less consistent with timing stuff.

SethMichaelLarson commented:
Looks like one of the `SelectSelector` test case is failing on Travis Linux builds but all other selectors are fine. :) I will see what I can do about this after the Mac OS builds also finish.

SethMichaelLarson commented:
@Lukasa Thoughts on making `TOLERANCE` a relative value rather than an absolute value?

SethMichaelLarson commented:
So it's looking like the CI will sometimes just take absurdly long in some circumstances.  (Like [this job](https://travis-ci.org/shazow/urllib3/jobs/170861263) which randomly took 50% longer than expected on a random `PollSelector` test case.  Perhaps we can increase tolerance values if the use of a CI is detected? This is what I was afraid of with testing upper and lower on the same value while using the CI. :/

Lukasa commented:
Yeah, you can use an environment variable to detect Travis and adjust the values accordingly.

SethMichaelLarson commented:
@Lukasa Okay, I will do this, it might be a larger boost but it should only affect upper-bounds on time requirements.

SethMichaelLarson commented:
@Lukasa One of the benefits of this is we can tightly constrain tests on our own machines while still accepting fuzzy timing values for CIs. I'll have to add another check eventually once we merge #1006 because I'm sure that AppVeyor is not much better.

SethMichaelLarson commented:
I'm checking for `"TRAVIS"` being in `os.environ` as suggested by Travis but it doesn't seem to be changing the values of `TOLERANCE` as I've told it to.... any suggestions on why it's not picking up that the tests are being run on Travis, @Lukasa ?

Lukasa commented:
Ah, `tox` doesn't usually pass environment variables though. You'll need to expand [this list](https://github.com/shazow/urllib3/blob/master/tox.ini#L11).

SethMichaelLarson commented:
@Lukasa Oh! I don't know much about `tox` so this is a good learning opportunity. :) It seems like a very powerful tool.

SethMichaelLarson commented:
Mac OS is so slow! :/ Mac OS + Python 3.5 took __over twice as long__ on a test than it should have that all other Mac OS and Linux passed.  Could you restart that build, @Lukasa ? :pray:

SethMichaelLarson commented:
Thanks @Lukasa :) Looks like we're passing now and I'm ready for another round of review.

Lukasa commented:
Ok, I probably won't get to this until tomorrow. :smile:

SethMichaelLarson commented:
@Lukasa Sounds good to me, see you at 7 AM :P

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,672 @@
+        if not hasattr(signal, "SIGALRM"):
+            self.skipTest("Platform doesn't have signal.SIGALRM")
+        if not hasattr(self, "alarm_thread"):
+            self.alarm_thread = None
```
So this should be avoidable by having a class property called `alarm_thread` that is initialized to `None`. Any writes to `self.alarm_thread` won't overwrite the class variable, they'll just give a more specific value for that instance.

SethMichaelLarson replied:
I'm sorry if I don't have this correct.  Is this what you meant by class variable?

```python
class AlarmMixin(object):
    alarm_thread = None  # This right here?

    def set_alarm(self, timeout):
        ...
```

I do not have a lot of experience with how Python class behaviours act in all these situations. :) A chance to learn!

Lukasa replied:
Yup, that's what I meant. =D

SethMichaelLarson replied:
Okay good! :) :tada:

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,672 @@
+        self.alarm_thread.start()
+
+    def cancel_alarm(self):
+        if getattr(self, "alarm_thread", None) is not None:
```
Changing the `alarm_thread` to a class member fixes this issue too.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,672 @@
+    def __enter__(self):
+        self.start_time = get_time()
+
+    def __exit__(self, *args):
```
Let's take `**kwargs` as well, just to be forward-looking.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,672 @@
+
+
+class TimerContext(object):
+    def __init__(self, testcase, lower=None, upper=None ):
```
Rogue trailing whitespace.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,672 @@
+@skipUnless(selectors.HAS_SELECT, "Platform doesn't have a selector")
+class WaitForIOTest(unittest.TestCase, AlarmMixin, TimerMixin):
+    """ Tests for the higher level wait_for_* functions. """
+    TOLERANCE = 0.05
```
This doesn't appear to be used for anything.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,672 @@
+            self.alarm_thread.join(0.0)
+        self.alarm_thread = None
+
+    def make_alarm(self, handler, duration):
```
This is a nit, and you can tell me I'm being an idiot if you want, but these arguments feel backwards to me. I'd rather have the duration first I think.

SethMichaelLarson replied:
Definitely not an idiot :P They didn't feel right either way to me so I picked a way and hoped someone had a stronger opinion than me. :) I will change it to be `timeout`, `handler` because now in my mind it is definitely the correct way. :) I think this is because `signal.signal` is `signum`, `handler` :)

SethMichaelLarson replied:
I made this change. :smile:

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,672 @@
+
+        self.make_alarm(lambda *args: wr.send(b'x'), SHORT_SELECT)
+
+        with self.assertTakesTime(lower=SHORT_SELECT, upper=LONG_SELECT):
```
Once again, presumably if this takes as long as `LONG_SELECT` this test should have failed.

SethMichaelLarson replied:
I changed to `upper=SHORT_SELECT`.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,672 @@
+class BaseSelectorTestCase(unittest.TestCase, AlarmMixin, TimerMixin):
+    """ Implements the tests that each type of selector must pass. """
+    SELECTOR = selectors.DefaultSelector
+    TOLERANCE = 0.05
```
Once again, as far as I can see this isn't being used for anything.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,672 @@
+        s = self.make_selector()
+        readers = []
+        writers = []
+        for i in range(32):
```
Given that we aren't using `i`, let's change it to `_`.

SethMichaelLarson replied:
I made this change. :)

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,672 @@
+        with self.assertTakesTime(upper=SHORT_SELECT):
+            self.assertEqual(1, len(s.select(timeout=None)))
+            self.assertEqual(1, len(s.select(timeout=None)))
+            self.assertEqual(1, len(s.select(timeout=None)))
```
Why are there three calls here? Might be good to add a comment if you think they're necessary.

SethMichaelLarson replied:
Three calls are not needed.  I reduced to one.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,672 @@
+            self.assertEqual(0, len(s.select(timeout=0)))
+
+        with self.assertTakesTime(lower=SHORT_SELECT):
+            self.assertEqual(0, len(s.select(timeout=SHORT_SELECT)))
```
Presumably we want an upper threshold too.

SethMichaelLarson replied:
Yes, I made this change.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,672 @@
+
+        self.make_alarm(lambda *args: None, SHORT_SELECT)
+
+        # The x1.1 is because SelectSelector is really slow on CI.
```
That's possible, but this test will run with the upper timeout on all selectors. Is there a better approach to this?

Lukasa replied:
I should note that in addition to the 1.1 there will be a further multiplication of 1.55.

SethMichaelLarson replied:
You are correct, I removed the * 1.1.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,672 @@
+
+        self.make_alarm(first_alarm, SHORT_SELECT)
+
+        with self.assertTakesTime(lower=SHORT_SELECT * 2, upper=SHORT_SELECT * 4):
```
Really? Is `*4` really needed?

SethMichaelLarson replied:
Probably not, I will try with * 3 because I suspect that alarm handling on the CI is what is so slow. :/

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,672 @@
+
+        def alarm_exception(*args):
+            err = OSError()
+            err.errno = 100
```
It'd be nice to give this number a name, because it looks a bit weird just floating out here as a random number.

SethMichaelLarson replied:
I decided to use `errno.EACESS` :) Is this okay?

Lukasa replied:
Should be fine. =)

SethMichaelLarson commented:
Thank you @Lukasa for the round of review. :) I am excited as I feel we are nearing completion. :) Do we already have approval from shazow about this large of a change?

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,664 @@
+
+        self.make_alarm(SHORT_SELECT, lambda *args: wr.send(b'x'))
+
+        with self.assertTakesTime(upper=SHORT_SELECT * 2):
```
I think in each place we put tolerance that is *beyond* the tolerance that is baked into this helper, we should really put a comment explaining why we think we had to do this.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,664 @@
+
+@skipUnless(hasattr(selectors, "KqueueSelector"), "Platform doesn't have a KqueueSelector")
+class KqueueSelectorTestCase(BaseSelectorTestCase, ScalableSelectorMixin):
+    SELECTOR = getattr(selectors, "KqueueSelector", None)
```
Given that all of these are protected by a `hasattr`, let's just use `selectors.WhateverSelector`.

SethMichaelLarson replied:
Good idea, I made this change.

Lukasa commented on urllib3/util/__init__.py:
```diff
@@ -24,6 +24,10 @@
+from .selectors import (
+    wait_for_read,
+    wait_for_write
+)
```
At this point I think we shouldn't bother doing this cross-import thing. Either we want our backport to maintain the same interface as the stdlib, in which case the implementation of these two functions should be moved fully into `util`, or we don't care, in which case our code should just import them directly from `selectors`. Up to you @SethMichaelLarson, but we should aim to do one of them.

SethMichaelLarson replied:
I think that having the same interface as traditional selectors is valuable. I will move the `wait_for_read` and `wait_for_write` to a different file.  Any suggestions on where these should end up? :)

Lukasa replied:
Maybe in a new file? `wait.py`, possibly?

SethMichaelLarson commented on test/test_selectors.py:
```diff
@@ -0,0 +1,664 @@
+    def set_alarm(self, timeout):
+        if not hasattr(signal, "SIGALRM"):
+            self.skipTest("Platform doesn't have signal.SIGALRM")
+        if self.alarm_thread is None:
```
I will remove it. :)

SethMichaelLarson commented on test/test_selectors.py:
```diff
@@ -0,0 +1,664 @@
+            self.alarm_thread.join(0.0)
+        self.alarm_thread = None
+
+    def make_alarm(self, duration, handler):
```
Sure thing :)

SethMichaelLarson commented on test/test_selectors.py:
```diff
@@ -0,0 +1,664 @@
+    def test_wait_timeout(self):
+        rd, wr = self.make_socketpair()
+        with self.assertTakesTime(lower=0.0, upper=0.1):
+            selectors.wait_for_read([rd], timeout=0.01)
```
You're right! I changed all the numerical constants in this test case to be named.

SethMichaelLarson commented on test/test_selectors.py:
```diff
@@ -0,0 +1,664 @@
+        with self.assertTakesTime(lower=0.0, upper=0.1):
+            selectors.wait_for_read([rd], timeout=0.01)
+
+    @skipUnless(hasattr(signal, "alarm"), "Platform doesn't have signal.alarm()")
```
Sure thing, this has been done. :)

SethMichaelLarson commented:
Quick question that I had, do you think it is valuable to make `SelectorError` a sub-class of OSError as it can only be raised by an OS, IO, or select.error? Or should it be it's own thing?

Lukasa commented:
Nah, don't bother doing the subclass. =D

SethMichaelLarson commented:
Sounds good to me. :)

SethMichaelLarson commented:
@Lukasa Travis has not been picking up any of my past commits. :/  Very strange behaviour from Travis ... seems to work for other PRs.

SethMichaelLarson commented:
Aha! Travis doesn't build PRs if there is a merge conflict! I will resolve this now so we can get our nice Travis builds back again. :)  [Referece](https://docs.travis-ci.com/user/pull-requests#My-Pull-Request-isn%E2%80%99t-being-built)

SethMichaelLarson commented:
Random fail on Mac OS + Python 3.4, took way longer than it should have.  Should `UPPER_TOLERANCE` be even higher for Mac OS + Travis? :/

shazow commented:
As @lukasa said, he has the final word right now. :)

A thought from the sidelines: Do we have any notion of how flakey these sleep-based tests are going to be? We spent a lot of effort migrating from sleep-based to event-based tests in our main test suite, it was pretty painful. For a while the tests were so flakey that people wouldn't bother to run them regularly (myself included). If we can't avoid flakey tests for this, might be another reason to keep this sub-package separate (in backports or whatever).

SethMichaelLarson commented:
The problem with the tests isn't with running locally, it's with running in the containers on Travis. The machines are very slow and don't respond as far as they should to system calls. Running locally the tests always pass with tons of time to spare on my old Linux machine and Windows machine.

Lukasa commented:
So let's ask the question: can we get by without having tests for the specific timeout logic? Or with having vastly fewer?

SethMichaelLarson commented:
We can get by, this is the reason why I was originally just doing single-bounded tests for timeouts, I knew that Travis was going to be really finicky :P The main thing was that I wanted to make sure that selectors behave correctly in all circumstances that they'll be subjected to having to deal with system interrupts and restarting, and it seems that they do from the Travis testing.

I'll leave this decision to you @Lukasa, I don't really know what is best here. We still want to maintain coverage unless we move this module to another uncovered place.

Lukasa commented:
Let's take a different direction and *exclude* all the timing tests from running on Travis. We can then remove the special-casing of allowing large tolerances and keep all the timing tests as precise as possible.

These tests can continue to run when the tests are run locally, which allows us to have *some* fallback for testers.

SethMichaelLarson commented:
@Lukasa That's the best direction to go IMO but I really dislike skipping tests :P That'll be a really easy change.

SethMichaelLarson commented:
Passed on CI, ready to review again. :)

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,673 @@
+LONG_SELECT = 0.2
+SHORT_SELECT = 0.01
+
+# Tolerance values for CI timer/speed fluctuations.
```
This comment should suggest that the tolerance is not for CI but for general timer/speed fluctuations.

SethMichaelLarson replied:
Sure I can do this.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,673 @@
+# Tolerance values for CI timer/speed fluctuations.
+TOLERANCE = 0.075
+
+# Travis CI detection, sometimes Travis is extremely slow.
```
This comment should elaborate that we often just skip tests outright on Travis.

SethMichaelLarson replied:
We don't skip tests per-say, we just skip verification points for timing. :) All tests still run.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,673 @@
+        total_time = self.end_time - self.start_time
+
+        # Skip timing on CI due to flakiness.
+        if not TRAVIS_CI:
```
Rather than this multiple indent, let's just have a short-circuit return.

SethMichaelLarson replied:
I will add this.

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,673 @@
+
+        try:
+            s.select(LONG_SELECT)
+            self.fail("select() didn't raise SelectorError")
```
Let's move this to an `else` block. =D

SethMichaelLarson commented:
Tests have passed, ready to re-review. :)

SethMichaelLarson commented:
Huzzah! :tada:

Lukasa commented:
Ok, I officially think I'm happy with this change.

I'm not going to merge this right away, though: it's a large change that needs some time to settle in the master branch. I'm currently planning to push a 1.19 release this week to ship some of the changes we've landed there and to fix up some install problems we introduced in 1.18. I'll merge after that. That should allow other contributors who are interested in this work to do some more reviewing as well.

In other news, well done @SethMichaelLarson! This is a big patch, but it frees us up to start doing some much smarter things with I/O management on many platforms. I'm really looking forward to the direction this should start taking urllib3 in. Congratulations!

SethMichaelLarson commented:
Yes!!! That sounds like a great plan. :) I can't wait to see where else besides Happy Eyeballs this will be used, I am very excited! Now I can continue work on #1000 proper. :P

shazow commented:
:custard: :sparkles: 🍍

Lukasa commented on test/test_selectors.py:
```diff
@@ -0,0 +1,678 @@
+SHORT_SELECT = 0.01
+
+# Tolerance values for timer/speed fluctuations.
+TOLERANCE = 0.075
```
Honestly that seems like a good first start. Potentially even a bit higher: 0.75, maybe. Defensiveness here is probably a good idea, especially as the tolerance required to swamp the differences in timing would be 20.

SethMichaelLarson replied:
Should we just go to 0.75 then, having these tests randomly fail when run locally will just confuse people as to what they broke haha

SethMichaelLarson commented:
Question @Lukasa would it mess anything up to merge this branch into the branch used by #1000?

Lukasa commented:
It would be better if you didn't. If you want the CI rig to run against it, the best plan is to make a third branch and merge both this and #1000 into it, and then open that up as a temp PR marked "[DO NOT MERGE]".

SethMichaelLarson commented:
@Lukasa Sure I just haven't ever needed to do that so I figured I'd ask before mucking up all the work I've done. :P I'll stick with the temporary selectors file I've been using on that branch until this is merged.

SethMichaelLarson commented:
@Lukasa Should other PRs and issues regarding this be closed now that this is finalized?  So far I've found #589, #685, #700, #998, and maybe #681

Lukasa commented:
Well identified. Sadly I have no evidence that #681 will be helped by this PR, so we can't close that one.

SethMichaelLarson commented:
@Lukasa Yeah I didn't know if this work would have any effect on that issue, I can try to recreate it and see what happens later.

SethMichaelLarson commented:
Will today be the big day? :tada:

Lukasa commented:
Once we can get a green test run, you bet it will!

SethMichaelLarson commented:
Looks like it's a green light! :)

SethMichaelLarson commented:
WHY MUST THAT BUTTON BE THERE???

Lukasa commented:
Just to punish you I'm afraid.

SethMichaelLarson commented:
Now it's done after my silly mistake. :)

Lukasa commented:
Hooray, merged!
